vunit/sim_if/xcelium.py

```diff
@@ -0,0 +1,694 @@
+        )
+        write_file(argsfile, "\n".join(args))
+        return [cmd, "-F", argsfile]
+
+    def compile_verilog_file_command(self, source_file):
+        """
+        Returns commands to compile a Verilog file
+        """
+        cmd = str(Path(self._prefix) / "xrun")
+        args = []
+        args += ["-compile"]
+        args += ["-nocopyright"]
+        args += ["-licqueue"]
+        # "Ignored unexpected semicolon following SystemVerilog description keyword (endfunction)."
+        args += ["-nowarn UEXPSC"]
+        # "cds.lib Invalid path"
+        args += ["-nowarn DLCPTH"]
+        # "cds.lib Invalid environment variable ''."
+        args += ["-nowarn DLCVAR"]
+        args += source_file.compile_options.get("xcelium.xrun_verilog_flags", [])
+        args += ['-cdslib "%s"' % self._cdslib]
+        args += self._hdlvar_args()
+        args += [
+            '-log "%s"'
+            % str(
+                Path(self._output_path)
+                / ("xrun_compile_verilog_file_%s.log" % source_file.library.name)
+            )
+        ]
+        if not self._log_level == "debug":
+            args += ["-quiet"]
+        else:
+            args += ["-messages"]
+            args += ["-libverbose"]
+        for include_dir in source_file.include_dirs:
+            args += ['-incdir "%s"' % include_dir]
+
+        # for "disciplines.vams" etc.
+        args += ['-incdir "%s/tools/spectre/etc/ahdl/"' % self._cds_root_xrun]
+
+        for key, value in source_file.defines.items():
+            args += ["-define %s=%s" % (key, value.replace('"', '\\"'))]
+        args += ['-xmlibdirname "%s"' % str(Path(source_file.library.directory).parent)]
+        args += ["-makelib %s" % source_file.library.name]
+        args += ['"%s"' % source_file.name]
+        args += ["-endlib"]
+        argsfile = str(
+            Path(self._output_path)
+            / ("xrun_compile_verilog_file_%s.args" % source_file.library.name)
+        )
+        write_file(argsfile, "\n".join(args))
+        return [cmd, "-F", argsfile]
+
+    def create_library(self, library_name, library_path, mapped_libraries=None):
+        """
+        Create and map a library_name to library_path
+        """
+        mapped_libraries = mapped_libraries if mapped_libraries is not None else {}
+
+        lpath = str(Path(library_path).resolve().parent)
+
+        if not file_exists(lpath):
+            os.makedirs(lpath)
+
+        if (
+            library_name in mapped_libraries
+            and mapped_libraries[library_name] == library_path
+        ):
+            return
+
+        cds = CDSFile.parse(self._cdslib)
+        cds[library_name] = library_path
+        cds.write(self._cdslib)
+
+    def _get_mapped_libraries(self):
+        """
+        Get mapped libraries from cds.lib file
+        """
+        cds = CDSFile.parse(self._cdslib)
+        return cds
+
+    def simulate(  # pylint: disable=too-many-locals, too-many-branches
```

Comment:
This whole function is really complex and confusing. I would break things up a bit or find a better way to do all the branching.